### PR TITLE
8263677: Improve Character.isLowerCase/isUpperCase lookups

### PR DESCRIPTION
**Pull request by cl4es**
This patch changes the otherLowercase / otherUppercase bits to be set if either the codepoint is of type LOWERCASE_LETTER and UPPERCASE_LETTER, or the Unicode Other_Lowercase / Other_Uppercase property is set. This simplifies the lookup in Character.isLowerCase/isUpperCase to a single table lookup, which appears to be healthy for performance.

I also took the opportunity to clean up the somewhat dated GenerateCharacter utility class.

Testing: tier1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263677](https://bugs.openjdk.java.net/browse/JDK-8263677): Improve Character.isLowerCase/isUpperCase lookups

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to 5dc70a952dff1a77dec5693ef2305fdd34a6f0f6
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3028/head:pull/3028`
`$ git checkout pull/3028`

To update a local copy of the PR:
`$ git checkout pull/3028`
`$ git pull https://git.openjdk.java.net/jdk pull/3028/head`

### CODE REVIEW

**Comment by cl4es** [comment]
Baseline:
```
Benchmark               (codePoint)  Mode  Cnt   Score   Error  Units
Characters.isLowerCase            9  avgt    5  13.809 ± 0.032  ns/op
Characters.isLowerCase           65  avgt    5  13.811 ± 0.052  ns/op
Characters.isLowerCase           97  avgt    5  12.552 ± 0.057  ns/op
Characters.isLowerCase          128  avgt    5  13.823 ± 0.076  ns/op
Characters.isLowerCase          170  avgt    5  13.811 ± 0.066  ns/op
Characters.isLowerCase          223  avgt    5  12.556 ± 0.058  ns/op
Characters.isLowerCase          410  avgt    5  19.466 ± 0.104  ns/op
Characters.isLowerCase          430  avgt    5  20.718 ± 0.100  ns/op
Characters.isUpperCase            9  avgt    5  12.556 ± 0.056  ns/op
Characters.isUpperCase           65  avgt    5  12.559 ± 0.067  ns/op
Characters.isUpperCase           97  avgt    5  12.555 ± 0.055  ns/op
Characters.isUpperCase          128  avgt    5  12.559 ± 0.060  ns/op
Characters.isUpperCase          170  avgt    5  12.556 ± 0.036  ns/op
Characters.isUpperCase          223  avgt    5  12.554 ± 0.055  ns/op
Characters.isUpperCase          410  avgt    5  20.722 ± 0.129  ns/op
Characters.isUpperCase          430  avgt    5  19.459 ± 0.091  ns/op
```

Patch:
```
Benchmark               (codePoint)  Mode  Cnt   Score   Error  Units
Characters.isLowerCase            9  avgt    5  12.556 ± 0.035  ns/op
Characters.isLowerCase           65  avgt    5  12.562 ± 0.073  ns/op
Characters.isLowerCase           97  avgt    5  12.551 ± 0.062  ns/op
Characters.isLowerCase          128  avgt    5  12.553 ± 0.039  ns/op
Characters.isLowerCase          170  avgt    5  12.554 ± 0.051  ns/op
Characters.isLowerCase          223  avgt    5  12.552 ± 0.035  ns/op
Characters.isLowerCase          410  avgt    5  18.833 ± 0.068  ns/op
Characters.isLowerCase          430  avgt    5  18.832 ± 0.074  ns/op
Characters.isUpperCase            9  avgt    5  12.555 ± 0.050  ns/op
Characters.isUpperCase           65  avgt    5  12.557 ± 0.041  ns/op
Characters.isUpperCase           97  avgt    5  12.554 ± 0.056  ns/op
Characters.isUpperCase          128  avgt    5  12.554 ± 0.055  ns/op
Characters.isUpperCase          170  avgt    5  12.555 ± 0.054  ns/op
Characters.isUpperCase          223  avgt    5  12.553 ± 0.036  ns/op
Characters.isUpperCase          410  avgt    5  18.831 ± 0.099  ns/op
Characters.isUpperCase          430  avgt    5  18.826 ± 0.047  ns/op
```

**Review by erikj79 — APPROVED** [review]
Looks good from build point of view. I like the code cleanups.

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 1]
```diff
@@ -314,7 +307,7 @@ static void FAIL(String s) {
     static long[] buildMap(UnicodeSpec[] data, SpecialCaseMap[] specialMaps, PropList propList)
     {
         long[] result;
-        if (bLatin1 == true) {
+        if (bLatin1) {
```
Comment (RogerRiggs) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:310]: perhaps shorten to:
final long[] result = new long[bLatin1 ? 256 : 1 << 16];

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 2]
```diff
@@ -649,18 +649,14 @@ static void addExProp(long[] map, PropList propList, String prop, long mask) {
             } // end MIDDLE
             // There is no block just like it already, so add it to
             // the buffer and put its index into the new map.
-            for (int k = 0; k < m; k++) {
-                buffer[ptr+k] = map[i+k];
-            }
+            if (m >= 0) System.arraycopy(map, i, buffer, ptr, m);
```
Comment (RogerRiggs) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:652]: If m == 0, you could skip the arraycopy.

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 3]
```diff
@@ -649,18 +649,14 @@ static void addExProp(long[] map, PropList propList, String prop, long mask) {
...
             newmap[i >> size] = (ptr >> size);
             ptr += m;
         } // end OUTER
         // Now we know how long the compressed table should be,
         // so create a new array and copy data from the temporary buffer.
         long[] newdata = new long[ptr];
-        for (int j = 0; j < ptr; j++) {
-            newdata[j] = buffer[j];
-        }
+        if (ptr >= 0) System.arraycopy(buffer, 0, newdata, 0, ptr);
```
Comment (RogerRiggs) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:659]: ditto, if ptr == 0, skip the arraycopy

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 4]
```diff
@@ -1688,8 +1584,8 @@ static String genAccess(String tbl, String var, int bits) {
 
     static void processArgs(String[] args) {
         StringBuffer desc = new StringBuffer("java GenerateCharacter");
-        for (int j=0; j<args.length; ++j) {
-            desc.append(" " + args[j]);
+        for (String arg : args) {
+            desc.append(" " + arg);
```
Comment (RogerRiggs) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:1588]: Avoid string concat:  
   desc.append(' ').append(arg);

**Review by magicus — APPROVED** [review]

**Review by naotoj — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]

**Pull request closed without merging**